fuel-core/src/executor.rs

```diff
@@ -103,51 +115,136 @@ impl Executor {
         let new_block_height = current_height + 1u32.into();
 
-        let mut block = FuelBlock {
-            header: FuelBlockHeader {
-                height: new_block_height,
-                parent_hash: current_hash,
-                time: Utc::now(),
-                da_height,
-                ..Default::default()
+        let block = PartialFuelBlock::new(
+            PartialFuelBlockHeader {
+                application: FuelApplicationHeader {
+                    // TODO: This should not be default
+                    da_height,
+                    generated: Default::default(),
+                },
+                consensus: FuelConsensusHeader {
+                    // TODO: This should not be default
+                    prev_root: Default::default(),
+                    height: new_block_height,
+                    time: Utc::now(),
+                    generated: Default::default(),
+                },
+                metadata: Default::default(),
             },
-            transactions: txs.into_iter().map(|t| t.as_ref().clone()).collect(),
-        };
+            txs.into_iter().map(|t| t.as_ref().clone()).collect(),
+        );
         // immediately execute block
-        self.execute(&mut block, ExecutionMode::Production).await?;
+        self.execute(ExecutionBlock::Production(block)).await?;
         Ok(())
     }
 
     #[tracing::instrument(skip(self))]
-    pub async fn execute(
+    pub async fn execute(&self, block: ExecutionBlock) -> Result<FuelBlock, Error> {
+        // Compute the block id before execution if there is one.
+        let pre_exec_block_id = block.id();
+        // Get the transaction root before execution if there is one.
+        let pre_exec_txs_root = block.txs_root();
+
+        // If there is full fuel block for validation then map it into
+        // a partial header.
+        let mut block = block.map_v(PartialFuelBlock::from);
+
+        // Create a new database transaction.
+        let mut block_db_transaction = self.database.transaction();
+
+        // Execute all transactions.
+        let execution_data = self
+            .execute_transactions(&mut block_db_transaction, block.as_mut())
+            .await?;
+
+        let ExecutionData {
+            coinbase,
+            message_ids,
+            mut tx_status,
+        } = execution_data;
+
+        // Now that the transactions have been executed, generate the full header.
+        let block = block.map(|b| b.generate(&message_ids[..]));
+
+        // check transaction commitment
+        if let Some(pre_exec_txs_root) = pre_exec_txs_root {
+            if block.header().transactions_root != pre_exec_txs_root {
+                return Err(Error::InvalidTransactionRoot)
+            }
+        }
+
+        let finalized_block_id = block.id();
+
+        debug!(
+            "Block {:#x} fees: {}",
+            pre_exec_block_id.unwrap_or_default(),
```

Comment:
Should this use the finalized id if the pre-exec block ID is none (e.g. production mode of partial header)?

```suggestion
            pre_exec_block_id.unwrap_or(finalized_block_id),
```